app/Main.hs

```diff
@@ -17,72 +18,52 @@ import           Data.Char
-checkBounds :: ((Int, Int), (Int, Int)) -> State Pong ()
+checkBounds :: ((Integer, Integer), (Integer, Integer)) -> State Pong ()
 checkBounds (xLim, yLim) = do
     p <- get
-    let x = view ballX p
-        y = view ballY p
-        vx = view ballDx p
-        vy = view ballDy p
+    let (x, y)   = p ^. ball . position
+        (vx, vy) = p ^. ball . velocity
         (x', dx') = boundary xLim (x, vx)
         (y', dy') = boundary yLim (y, vy)
 
-    ballX .= x'
-    ballY .= y'
-    ballDx .= dx'
-    ballDy .= dy'
+    (ball . position) .= (x', y')
+    (ball . velocity) .= (dx', dy')
 
-
-loop :: StateT Pong Curses ()
-loop = do
-    k <- lift checkKeyboard
+processInput :: KeyAction -> StateT Pong Curses ()
+processInput k = do
     pong <- get
-    let xLim = fromIntegral $ view playFieldWidth pong
-        yLim = fromIntegral $ view playFieldHeight pong
-    if k == Quit then lift quit
-       else do
-            case k of 
-                 Slower -> do
-                     let currentSpeed = view playFieldSpeed pong
-                         newSpeed = currentSpeed + 1000
-                     playFieldSpeed .= newSpeed
-                 Faster -> do
-                     let currentSpeed = view playFieldSpeed pong
-                         newSpeed = currentSpeed - 1000
-                     playFieldSpeed .= newSpeed
-                 RPaddleUp ->
-                     rightPaddleY -= movePaddle RPaddleUp pong 1
-                 RPaddleDn ->
-                     rightPaddleY += movePaddle RPaddleDn pong (yLim-6)
-                 LPaddleUp ->
-                     leftPaddleY -= movePaddle LPaddleUp pong 1
-                 LPaddleDn ->
-                     leftPaddleY += movePaddle LPaddleDn pong (yLim-6) 
-                 _ -> playFieldSpeed .= view playFieldSpeed pong  
-
-            hoist generalize $ checkBounds ((0, xLim - 1), (0, yLim - 1))
-
-            nextPong <- get
-            let x  = view ballX pong
-                y  = view ballY pong
-                x' = view ballX nextPong
-                y' = view ballY nextPong
-
-            lift $ updateDisplay (x, y) (x', y') pong nextPong
-            liftIO $ threadDelay $ view playFieldSpeed pong
+    let xLim = pong ^. playFieldSpecs . size . width
+        yLim = pong ^. playFieldSpecs . size . height
 
-            loop
-
-
-movePaddle :: KeyAction -> Pong -> Int -> Int
+    case k of
+        Slower ->
+            speed += 1000
+        Faster -> do
+            speed -= 1000
+        RPaddleUp ->
+            (rightPaddle . position . _y) -= movePaddle RPaddleUp pong 1
+        RPaddleDn ->
+            (rightPaddle . position . _y) += movePaddle RPaddleDn pong (yLim-6)
+        LPaddleUp ->
+            (leftPaddle . position . _y)  -= movePaddle LPaddleUp pong 1
+        LPaddleDn ->
+            (leftPaddle . position . _y)  += movePaddle LPaddleDn pong (yLim-6)
+        _ -> return ()
+
+    hoist generalize $ checkBounds ((0, xLim - 1), (0, yLim - 1))
```

Comment:
We can possibly get rid of `hoist generalize` and the `mmorph` dependency altogether if the type of `checkBounds` changes to `((Integer, Integer), (Integer, Integer)) -> StateT Pong Curses ()`

Reply:
I've opened another MR in case you want to see that change.